projects/packages/woocommerce-analytics/src/API/class-wc-analytics-tracking-proxy.php

```diff
@@ -0,0 +1,138 @@
+<?php
+/**
+ * REST API: WC_Analytics_Tracking_Proxy class
+ *
+ * @package automattic/woocommerce-analytics
+ */
+
+namespace Automattic\Woocommerce_Analytics;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class to handle tracking events via the REST API
+ *
+ * @since 0.7.0
+ */
+class WC_Analytics_Tracking_Proxy extends \WC_REST_Controller {
+
+	/**
+	 * Endpoint namespace.
+	 *
+	 * @var string
+	 */
+	protected $namespace = 'woocommerce-analytics/v1';
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'track';
+
+	/**
+	 * Register the routes for tracking.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => \WP_REST_Server::CREATABLE,
+					'callback'            => array( $this, 'track_events' ),
+					'permission_callback' => '__return_true', // no need to check permissions
+					'schema'              => array( $this, 'get_public_item_schema' ),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Track events.
+	 *
+	 * @param \WP_REST_Request $request Full data about the request.
+	 * @return \WP_REST_Response|\WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function track_events( $request ) {
+		$events = $request->get_json_params();
+
+		if ( ! is_array( $events ) || ( isset( $events['event_name'] ) ) ) {
+			// If $events is a single event (associative array), wrap it in an array.
+			$events = array( $events );
+		}
+
+		$results    = array();
+		$has_errors = false;
+
+		foreach ( $events as $index => $event ) {
```

Comment:
Feels it would be much clearer if we abstract the logic a bit by abstract a layer for processing a single event, which returns a status which then be packed and returned to the user.

Reply:
Great suggestion, @kangzj! I'll try refactoring the code in a follow-up PR. Thank you for your feedback 🙌